File: composeApp/src/iosMain/kotlin/org/example/kmpsamples/presentation/permissions/PlatformPermissionMapper.kt
package org.example.kmpsamples.presentation.permissions

import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType.PlatformMultiplePermissions
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType.PlatformSinglePermission

object PlatformPermissionMapper {
    fun toPlatformSinglePermissions(platformPermissionTypes: List<PlatformPermissionType>): List<PlatformSinglePermission> {
        return platformPermissionTypes.flatMap { permissionType ->
            when (permissionType) {
                is PlatformSinglePermission -> listOf(
                    permissionType
                )

                is PlatformMultiplePermissions -> permissionType.permissions
            }
        }
    }
}

File: iosApp/iosApp/Permissions/PlatformPermissionMapper.swift
import Foundation
import ComposeApp

struct PlatformPermissionMapper {
    static func toPlatformSinglePermissions(_ platformPermissionTypes: [PlatformPermissionType]) -> [PlatformPermissionTypePlatformSinglePermission] {
        
        return platformPermissionTypes.flatMap { permissionType in
            switch permissionType {
            case let singlePermission as PlatformPermissionTypePlatformSinglePermission:
                return [singlePermission]
            case let multiplePermissions as PlatformPermissionTypePlatformMultiplePermissions:
                return multiplePermissions.permissions
            default:
                fatalError("Not supported platform permission type: \(permissionType)")
            }
        }
    }
}
